src/Core/Signers/SignerV4.php

```diff
@@ -123,89 +123,34 @@ public function sign(Request $request, ?Credentials $credentials): void
         $request->setHeader('authorization', $authorizationHeader);
     }
 
-    private function getContentLength(Request $request): ?int
+    private function prepareBody(Request $request, string $amzDate, string $credentialScope, string &$signature, string $signingKey): void
     {
-        if ($request->hasHeader('content-length')) {
-            return (int) ((array) $request->getHeader('content-length'))[0];
-        }
-
         $body = $request->getBody();
-        if (\is_string($body)) {
-            return \strlen($body);
-        }
 
-        if (\is_resource($body)) {
-            return fstat($body)['size'] ?? null;
+        if ($request->hasHeader('content-length')) {
+            $contentLength = (int) ((array) $request->getHeader('content-length'))[0];
+        } else {
+            $contentLength = $body->length();
         }
 
-        return null;
-    }
-
-    private function prepareBody(Request $request, string $amzDate, string $credentialScope, string &$signature, string $signingKey): void
-    {
-        $body = $request->getBody();
-
-        // we can't manage signature of undefined length closure
-        $contentLength = $this->getContentLength($request);
+        // we can't manage signature of undefined length. Let's convert it to string
         if (null === $contentLength) {
-            if (\is_callable($body)) {
-                $buffer = '';
-                while (true) {
-                    if (!\is_string($data = $body(self::CHUNK_SIZE))) {
-                        throw new InvalidArgument(sprintf('The return value of the "body" option callback must be a string, %s returned.', \gettype($data)));
-                    }
-
-                    if ('' == $data) {
-                        break;
-                    }
-
-                    $buffer .= $data;
-                }
-
-                $request->setBody($body = $buffer);
-            } elseif (\is_resource($body)) {
-                $request->setBody($body = \stream_get_contents($body));
-            }
+            $request->setBody($body = new StringStream($body->stringify()));
+            $contentLength = $body->length();
         }
 
-        if (\is_string($body)) {
-            $request->setHeader('x-amz-content-sha256', hash('sha256', $body));
+        // no need to stream small body
+        if ($contentLength < self::CHUNK_SIZE) {
+            if (!$body instanceof StringStream) {
+                $request->setBody($body = new StringStream($body->stringify()));
+            }
```

Comment:
These 3 lines could be removed if we were using a named factory. Either: 

```php
$body = StreamFactory::stringStream($body);
```
or
```php
$body = StringStream::create($body);
```

Reply:
good idea.